File: web/src/api/sql.ts
import i18n from "@/lib/i18n"
import { post, request } from "@/api"
import type {
  ColumnFilter,
  DBSchema,
  DBTables,
  GenSQLKind,
  ObjectDDLResult,
  ObjectDDLType,
  ObjectNode,
  SortSpec,
  SQLAuditEntry,
  SQLExecMode,
  SQLFavorite,
  SQLHistoryItem,
  SQLPingResult,
  SQLQueryRequest,
  SQLQueryResult,
  TableDataRequest,
  TableDataResult,
  WorkspaceState,
} from "@/types"

// ---- SQL 查询终端 ----

// 统一执行（Navicat 式）：支持多语句批量执行 + 选中执行，返回结果集数组。
// 写操作确认由调用方（store）在发送前完成。
export const runSql = (payload: SQLQueryRequest) =>
  post<SQLQueryResult[]>("/api/sql/run", payload)

export const pingConnection = (connId: string) =>
  post<SQLPingResult>("/api/sql/ping", { connId })

export const fetchSqlHistory = (connId: string) =>
  request<SQLHistoryItem[]>(`/api/sql/history?connId=${encodeURIComponent(connId)}`)

export const clearSqlHistory = (connId: string) =>
  request<{ ok: boolean }>(`/api/sql/history?connId=${encodeURIComponent(connId)}`, { method: "DELETE" })

export const listFavorites = () =>
  request<SQLFavorite[]>(`/api/sql/favorites`, { method: "GET" })

export const addFavorite = (connId: string, params: {
  sql: string
  db?: string
  mode?: SQLExecMode
  title?: string
}) =>
  post<{ ok: boolean; id: string }>(`/api/sql/favorites?connId=${encodeURIComponent(connId)}`, {
    connId,
    ...params,
  })

export const deleteFavorite = (id: string) =>
  request<{ ok: boolean }>(`/api/sql/favorites?id=${encodeURIComponent(id)}`, { method: "DELETE" })

export const renameFavorite = (id: string, title: string) =>
  post<{ ok: boolean }>(`/api/sql/favorites`, {
    id,
    title,
  })

// 审计日志（只读）：分页读取，按连接过滤；connId 为空 = 全部连接
export const fetchSqlAudit = (connId: string, limit = 100, offset = 0) =>
  request<SQLAuditEntry[]>(
    `/api/sql/audit?connId=${encodeURIComponent(connId)}&limit=${limit}&offset=${offset}`,
  )

// ---- 查询工作区（后端持久化，按连接） ----

export const fetchWorkspace = (connId: string) =>
  request<WorkspaceState>(`/api/sql/workspace?connId=${encodeURIComponent(connId)}`)

export const saveWorkspace = (connId: string, state: WorkspaceState) =>
  request<{ ok: boolean }>(`/api/sql/workspace`, {
    method: "PUT",
    body: JSON.stringify({
      connId,
      tabs: state.tabs,
      activeId: state.activeId,
      tabSettings: state.tabSettings,
    }),
  })

// 获取对象创建语句（表/视图/函数/存储过程 DDL）
export const getObjectDDL = (connId: string, db: string, type: ObjectDDLType, name: string) =>
  request<ObjectDDLResult>(
    `/api/sql/ddl?connId=${encodeURIComponent(connId)}&db=${encodeURIComponent(db)}&type=${encodeURIComponent(type)}&name=${encodeURIComponent(name)}`,
  )

// ---- 数据浏览器 / 对象树（复用连接表树接口） ----

// 从连接表树接口转换为对象树。
// 后端 objects 的 key 为 _views/_functions/_procedures（带下划线前缀，见 engine/objects.go objectKindDirs）。
// PG 系返回 schemas 分层（库 → schema → 分组 → 对象），叶子名称带 "schema." 前缀，
// 打开对象时可直接作为限定名传给后端；MySQL/Oracle 保持 库 → 分组 → 对象 结构。
export const buildObjectTree = (databases: { name: string; tables: string[]; objects?: Record<string, string[]>; schemas?: DBSchema[] }[]): ObjectNode[] =>
  databases.map((db) => ({
    name: db.name,
    type: "db",
    children: db.schemas?.length
      ? db.schemas.map((s) => ({
          name: s.name,
          type: "schema" as const,
          children: [
            ...(s.tables ?? []).map((t) => ({ name: `${s.name}.${t}`, type: "table" as const })),
            ...(s.objects?.["_views"] ?? []).map((v) => ({ name: `${s.name}.${v}`, type: "view" as const })),
            ...(s.objects?.["_functions"] ?? []).map((f) => ({ name: `${s.name}.${f}`, type: "function" as const })),
            ...(s.objects?.["_procedures"] ?? []).map((p) => ({ name: `${s.name}.${p}`, type: "procedure" as const })),
          ],
        }))
      : [
          ...(db.tables ?? []).map((t) => ({ name: t, type: "table" as const })),
          ...(db.objects?.["_views"] ?? []).map((v) => ({ name: v, type: "view" as const })),
          ...(db.objects?.["_functions"] ?? []).map((f) => ({ name: f, type: "function" as const })),
          ...(db.objects?.["_procedures"] ?? []).map((p) => ({ name: p, type: "procedure" as const })),
        ],
  }))

// ---- 对象树分级加载构建辅助 ----

// buildDbStub 未加载的库节点（灰显，点击加载）
export const buildDbStub = (name: string): ObjectNode => ({ name, type: "db", loaded: false })

// buildSchemaStub 未加载的 schema 节点（灰显 + 表计数，点击加载）
export const buildSchemaStub = (name: string, tableCount: number): ObjectNode => ({
  name,
  type: "schema",
  count: tableCount,
  loaded: false,
})

// buildSchemaLeafs 已加载 schema 的叶子节点（表/视图/函数/过程，限定名）
export const buildSchemaLeafs = (s: DBSchema): ObjectNode[] => [
  ...(s.tables ?? []).map((t) => ({ name: `${s.name}.${t}`, type: "table" as const })),
  ...(s.objects?.["_views"] ?? []).map((v) => ({ name: `${s.name}.${v}`, type: "view" as const })),
  ...(s.objects?.["_functions"] ?? []).map((f) => ({ name: `${s.name}.${f}`, type: "function" as const })),
  ...(s.objects?.["_procedures"] ?? []).map((p) => ({ name: `${s.name}.${p}`, type: "procedure" as const })),
]

// buildDbLeafs 已加载库的非 PG 叶子节点（MySQL/Oracle 无 schema 层，裸名）
export const buildDbLeafs = (db: DBTables): ObjectNode[] => [
  ...(db.tables ?? []).map((t) => ({ name: t, type: "table" as const })),
  ...(db.objects?.["_views"] ?? []).map((v) => ({ name: v, type: "view" as const })),
  ...(db.objects?.["_functions"] ?? []).map((f) => ({ name: f, type: "function" as const })),
  ...(db.objects?.["_procedures"] ?? []).map((p) => ({ name: p, type: "procedure" as const })),
]

// buildDbNode 已加载库节点：PG 系 children 为 schema 节点；MySQL/Oracle 为分组叶子
export const buildDbNode = (db: DBTables): ObjectNode => ({
  name: db.name,
  type: "db",
  loaded: true,
  children: db.schemas?.length
    ? db.schemas.map((s) => ({ name: s.name, type: "schema" as const, loaded: true, children: buildSchemaLeafs(s) }))
    : buildDbLeafs(db),
})

// 查询表数据：走专用分页接口，一次返回当前页数据与全表总行数（total）。
// 后端内部用 cydb 的 SQLStmt.Count 计算总数，不写审计/历史，避免二次 COUNT 的审计冗余。
export const fetchTableData = async (req: TableDataRequest): Promise<TableDataResult> => {
  const result = await post<TableDataResult>("/api/sql/table", {
    connId: req.connId,
    db: req.db,
    table: req.table,
    page: req.page,
    pageSize: req.pageSize,
    sortSpecs: req.sortSpecs,
    excludeColumns: req.excludeColumns,
    filters: req.filters,
  })
  return result
}

// 表数据导出 Excel：直接 fetch 返回文件流（响应非 JSON，不走 request 封装）。
// 成功时触发浏览器下载；失败时解析 cygin 错误响应抛错。
export const exportTableExcel = async (req: TableDataRequest, maxRows = 100000): Promise<void> => {
  const authToken = sessionStorage.getItem("dbx_token") || "" // 与 api/index.ts 的 resolveToken 保持一致
  const res = await fetch("/api/sql/table-export", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...(authToken ? { "X-Auth-Token": authToken } : {}),
    },
    body: JSON.stringify({
      connId: req.connId,
      db: req.db,
      table: req.table,
      sortSpecs: req.sortSpecs,
      filters: req.filters,
      maxRows,
    }),
  })
  if (!res.ok) {
    // 尝试解析错误响应
    try {
      const b = await res.json()
      const detail = (b?.details ?? []).filter(Boolean).join("；")
      throw new Error(detail || b?.msg || i18n.t("api.exportFailed", { status: res.status }))
    } catch (e) {
      if (e instanceof Error && e.message) throw e
      throw new Error(i18n.t("api.exportFailed", { status: res.status }))
    }
  }
  const blob = await res.blob()
  const url = URL.createObjectURL(blob)
  const a = document.createElement("a")
  a.href = url
  a.download = `${req.table}.xlsx`
  document.body.appendChild(a)
  a.click()
  document.body.removeChild(a)
  URL.revokeObjectURL(url)
}

// 表浏览大字段单元格取值请求（按主键 + 列名定位单行单列）
export interface CellValuePayload {
  connId: string
  db: string
  table: string
  column: string
  pkColumns: string[]
  pkValues: unknown[]
}

// 获取单个大字段单元格的完整值（懒加载）
export const fetchTableCellValue = (payload: CellValuePayload) =>
  post<{ value: unknown }>("/api/sql/cell-value", payload)

// 表浏览单元格更新请求（named bind 更新）
export interface UpdateCellPayload {
  connId: string
  db: string
  table: string
  column: string
  value: unknown
  pkColumns: string[]
  pkValues: unknown[]
}

// 更新表浏览中的单个单元格，返回影响行数
export const updateTableCell = (payload: UpdateCellPayload) =>
  post<{ affectedRows: number }>("/api/sql/cell", payload)

// 表浏览整行删除请求（按主键定位，支持批量）
export interface DeleteRowsPayload {
  connId: string
  db: string
  table: string
  pkColumns: string[]
  rows: unknown[][] // 每行主键值数组（与 pkColumns 顺序一致）
}

// 删除表浏览中选中的整行（批量），返回累计影响行数
export const deleteTableRows = (payload: DeleteRowsPayload) =>
  post<{ affectedRows: number }>("/api/sql/delete-rows", payload)

// 表浏览新增行请求（用户显式填写的列与值）
export interface InsertRowPayload {
  connId: string
  db: string
  table: string
  columns: string[]
  values: unknown[]
}

// 表浏览新增一行（INSERT），返回影响行数
export const insertTableRow = (payload: InsertRowPayload) =>
  post<{ affectedRows: number }>("/api/sql/insert-row", payload)

// 表浏览快速生成 SQL 请求（右键菜单 → 后端按方言生成可执行语句，生成不执行）
export interface GenSQLPayload {
  connId: string
  db: string
  table: string
  kind: GenSQLKind
  columns: string[]
  rows: unknown[][] // 行数据（insert 支持多行；whereCell 时 [[值]]）
  pkColumns?: string[] // 主键列名（update/delete/selectByPk）
  skipColumns?: string[] // 跳过的列（insert 的自增列）
  filters?: ColumnFilter[] // 过滤条件（selectByFilter）
  sortSpecs?: SortSpec[] // 排序（selectByFilter）
}

// 快速生成 SQL 文本（方言正确的可执行语句，多条以分号换行分隔）
export const generateSQL = (payload: GenSQLPayload) =>
  post<{ sql: string }>("/api/sql/generate", payload)

File: internal/web/sql.go
// SQL 查询终端 API：查询 / 写操作 / 历史。
package web

import (
	"dqex/internal/engine"
	"dqex/internal/service"

	"github.com/gin-gonic/gin"
	"gitlab.mycyclone.com/rpa-platform/pk-infrakit-g/pkg/cygin"
)

// SQLRunReq SQL 执行请求（统一查询与写操作，支持多语句批量执行）
type SQLRunReq struct {
	ConnID string `json:"connId" binding:"required"`
	DB     string `json:"db"` // 目标库名（点对象树查表时传入，覆盖连接默认库）
	SQL    string `json:"sql" binding:"required"`
	Limit  int    `json:"limit"`  // 返回行数上限，<=0 走默认（1000），最大 1000
	Offset int    `json:"offset"` // 分页偏移
	Mask   bool   `json:"mask"`   // 结果集脱敏：敏感列（password/token/secret 等）统一打码
	Mode   string `json:"mode"`   // 执行模式：transform（默认，转换+限制）| raw（原始直传）
}

// handleSQLRun 统一执行 SQL（Navicat 式）：按分号分割多语句，逐条判断读写并执行，
// 返回结果集数组。写操作的安全确认由前端在发送前完成。
func handleSQLRun(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req SQLRunReq) ([]*engine.SQLQueryResult, error) {
		mode := req.Mode
		if mode == "" {
			mode = "transform"
		}
		results, err := svc.RunSQLScript(c.Request.Context(), req.ConnID, req.DB, req.SQL, req.Limit, req.Offset, mode)
		if err != nil {
			return nil, renderErr(c, err)
		}
		if req.Mask {
			for _, r := range results {
				engine.MaskResult(r)
			}
		}
		return results, nil
	})
}

// TablePageReq 对象树数据浏览分页请求：一次返回当前页数据与全表总行数。
// 独立于 /query（查询终端），是系统自动生成的浏览查询，不写审计/历史。
type TablePageReq struct {
	ConnID         string                `json:"connId" binding:"required"`
	DB             string                `json:"db"` // 目标库名（点对象树查表时传入，覆盖连接默认库）
	Table          string                `json:"table" binding:"required"`
	Page           int                   `json:"page"`
	PageSize       int                   `json:"pageSize"`
	SortSpecs      []engine.SortSpec     `json:"sortSpecs"`      // 多列排序（按顺序叠加 ORDER BY）
	ExcludeColumns []string              `json:"excludeColumns"` // 省略的大字段列名（二进制/超长文本，列表不取真实值）
	Filters        []engine.ColumnFilter `json:"filters"`        // 列过滤条件（AND 叠加）
	MaxRows        int                   `json:"maxRows"`        // 导出时行数上限（仅 /table-export 使用，默认 100000）
}

// handleSQLTablePage 对象树数据浏览：分页查表数据 + 全表总行数一次返回。
func handleSQLTablePage(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req TablePageReq) (*engine.TablePageResult, error) {
		return svc.QueryTablePage(c.Request.Context(), req.ConnID, req.DB, req.Table, req.Page, req.PageSize, req.SortSpecs, req.ExcludeColumns, req.Filters)
	})
}

// UpdateCellReq 表浏览单元格更新请求（named bind 更新）
type UpdateCellReq struct {
	ConnID    string   `json:"connId" binding:"required"`
	DB        string   `json:"db"`                           // 目标库名
	Table     string   `json:"table" binding:"required"`     // 表名
	Column    string   `json:"column" binding:"required"`    // 目标列名
	Value     any      `json:"value"`                        // 新值（null 表示 SET NULL）
	PKColumns []string `json:"pkColumns" binding:"required"` // 主键列名
	PKValues  []any    `json:"pkValues" binding:"required"`  // 主键值
}

// handleUpdateCell 更新表浏览中的单个单元格，返回影响行数。
func handleUpdateCell(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req UpdateCellReq) (map[string]any, error) {
		affected, err := svc.UpdateTableCell(c.Request.Context(), req.ConnID, req.DB, engine.UpdateCellParams{
			Table:     req.Table,
			SetColumn: req.Column,
			SetValue:  req.Value,
			PKColumns: req.PKColumns,
			PKValues:  req.PKValues,
		})
		if err != nil {
			return nil, renderErr(c, err)
		}
		return map[string]any{"affectedRows": affected}, nil
	})
}

// DeleteRowsReq 整行删除请求（按主键定位，支持批量删除）
type DeleteRowsReq struct {
	ConnID    string   `json:"connId" binding:"required"`
	DB        string   `json:"db"`                           // 目标库名
	Table     string   `json:"table" binding:"required"`     // 表名
	PKColumns []string `json:"pkColumns" binding:"required"` // 主键列名
	Rows      [][]any  `json:"rows" binding:"required"`      // 每行主键值数组（与 PKColumns 顺序一致）
}

// handleDeleteRows 删除表浏览中选中的整行（按主键定位，支持批量），返回累计影响行数。
func handleDeleteRows(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req DeleteRowsReq) (map[string]any, error) {
		affected, err := svc.DeleteTableRows(c.Request.Context(), req.ConnID, req.DB, req.Table, req.PKColumns, req.Rows)
		if err != nil {
			return nil, renderErr(c, err)
		}
		return map[string]any{"affectedRows": affected}, nil
	})
}

// InsertRowReq 新增行请求（用户显式填写的列与值）
type InsertRowReq struct {
	ConnID  string   `json:"connId" binding:"required"`
	DB      string   `json:"db"`                         // 目标库名
	Table   string   `json:"table" binding:"required"`   // 表名
	Columns []string `json:"columns" binding:"required"` // 写入列名（与 Values 顺序一致）
	Values  []any    `json:"values" binding:"required"`  // 对应列值（null 表示 NULL）
}

// handleInsertRow 表浏览视图新增一行（INSERT），返回影响行数。
func handleInsertRow(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req InsertRowReq) (map[string]any, error) {
		affected, err := svc.InsertTableRow(c.Request.Context(), req.ConnID, req.DB, engine.InsertRowParams{
			Table:   req.Table,
			Columns: req.Columns,
			Values:  req.Values,
		})
		if err != nil {
			return nil, renderErr(c, err)
		}
		return map[string]any{"affectedRows": affected}, nil
	})
}

// CellValueReq 单个大字段单元格取值请求：按主键 + 列名定位单行单列，返回完整值。
// 用于列表省略的大字段（二进制/超长文本）点击查看，避免列表一次传输大量数据。
type CellValueReq struct {
	ConnID    string   `json:"connId" binding:"required"`
	DB        string   `json:"db"`                           // 目标库名
	Table     string   `json:"table" binding:"required"`     // 表名
	Column    string   `json:"column" binding:"required"`    // 目标列名
	PKColumns []string `json:"pkColumns" binding:"required"` // 主键列名
	PKValues  []any    `json:"pkValues" binding:"required"`  // 主键值
}

// handleCellValue 返回单行单列的完整值（大字段懒加载）。
func handleCellValue(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req CellValueReq) (map[string]any, error) {
		val, err := svc.GetCellValue(c.Request.Context(), req.ConnID, req.DB, req.Table, req.Column, req.PKColumns, req.PKValues)
		if err != nil {
			return nil, renderErr(c, err)
		}
		return map[string]any{"value": val}, nil
	})
}

// SQLHistoryReq 历史记录请求
type SQLHistoryReq struct {
	ConnID string `query:"connId" binding:"required"`
}

func handleSQLHistory(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req SQLHistoryReq) ([]service.SQLHistoryItem, error) {
		return svc.SQLHistory(req.ConnID), nil
	})
}

func handleClearSQLHistory(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req SQLHistoryReq) (any, error) {
		svc.ClearSQLHistory(req.ConnID)
		return gin.H{"ok": true}, nil
	})
}

// FavoriteReq 收藏列表请求（全局共享，无需连接过滤；保留 ConnID 仅作来源标记上下文）
type FavoriteReq struct {
	ConnID string `query:"connId"`
}

// FavoriteAddReq 新增收藏请求
type FavoriteAddReq struct {
	ConnID string `json:"connId" binding:"required"`
	Title  string `json:"title"`
	DB     string `json:"db"`
	Mode   string `json:"mode"`
	SQL    string `json:"sql" binding:"required"`
}

// FavoriteRenameReq 重命名收藏请求
type FavoriteRenameReq struct {
	ID    string `json:"id" binding:"required"`
	Title string `json:"title" binding:"required"`
}

func handleListFavorites(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req FavoriteReq) ([]*service.SQLFavorite, error) {
		return svc.ListFavorites(), nil
	})
}

func handleAddFavorite(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req FavoriteAddReq) (any, error) {
		f := &service.SQLFavorite{
			ConnID: req.ConnID, // 仅作来源标记，便于跨连接回填时提示
			Title:  req.Title,
			DB:     req.DB,
			Mode:   req.Mode,
			SQL:    req.SQL,
		}
		if err := svc.AddFavorite(f); err != nil {
			return nil, renderErr(c, err)
		}
		return gin.H{"ok": true, "id": f.ID}, nil
	})
}

func handleDeleteFavorite(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req FavoriteReq) (any, error) {
		id := c.Query("id")
		if id == "" {
			return nil, cygin.NewError(cygin.ErrParamsInvalid)
		}
		if err := svc.DeleteFavorite(id); err != nil {
			return nil, renderErr(c, err)
		}
		return gin.H{"ok": true}, nil
	})
}

func handleRenameFavorite(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req FavoriteRenameReq) (any, error) {
		if err := svc.RenameFavorite(req.ID, req.Title); err != nil {
			return nil, renderErr(c, err)
		}
		return gin.H{"ok": true}, nil
	})
}

// SQLAuditReq 审计日志查询请求（分页，按连接过滤）
type SQLAuditReq struct {
	ConnID string `query:"connId"` // 空 = 全部连接
	Limit  int    `query:"limit"`  // <=0 默认 100，上限 500
	Offset int    `query:"offset"`
}

func handleSQLAudit(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req SQLAuditReq) ([]service.SQLAuditEntry, error) {
		return svc.SQLAudit(req.ConnID, req.Limit, req.Offset)
	})
}

// SQLPingReq 连接健康检测请求
type SQLPingReq struct {
	ConnID string `json:"connId" binding:"required"`
}

// PingResult 连接健康检测结果
type PingResult struct {
	OK        bool   `json:"ok"`
	ElapsedMS int64  `json:"elapsedMs"`
	Error     string `json:"error,omitempty"`
}

func handleSQLPing(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req SQLPingReq) (PingResult, error) {
		elapsed, err := svc.PingConnection(c.Request.Context(), req.ConnID)
		if err != nil {
			return PingResult{OK: false, Error: err.Error()}, nil
		}
		return PingResult{OK: true, ElapsedMS: elapsed}, nil
	})
}

// SQLGenReq 快速生成 SQL 请求（表浏览右键生成，字段平铺自 GenSQLParams）
type SQLGenReq struct {
	ConnID string `json:"connId" binding:"required"`
	DB     string `json:"db"` // 目标库名（点对象树查表时传入，覆盖连接默认库）
	engine.GenSQLParams
}

// handleSQLGen 快速生成 SQL 文本：行/单元格/过滤条件 → 方言正确的可执行语句，
// 供表浏览右键菜单生成预览（生成不执行，不写审计）。
func handleSQLGen(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req SQLGenReq) (map[string]string, error) {
		sql, err := svc.GenerateSQL(c.Request.Context(), req.ConnID, req.DB, req.GenSQLParams)
		if err != nil {
			return nil, renderErr(c, err)
		}
		return map[string]string{"sql": sql}, nil
	})
}

// SQLDDLReq 对象创建语句查询请求
type SQLDDLReq struct {
	ConnID string `query:"connId" binding:"required"`
	DB     string `query:"db"`                      // 库名（Oracle 下为 schema）
	Type   string `query:"type" binding:"required"` // table / view / function / procedure
	Name   string `query:"name" binding:"required"`
}

func handleSQLDDL(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req SQLDDLReq) (*service.ObjectDDLResult, error) {
		return svc.GetObjectDDL(c.Request.Context(), req.ConnID, req.DB, req.Type, req.Name)
	})
}

// WorkspaceGetReq 查询工作区读取请求（connId 走 query 参数）
type WorkspaceGetReq struct {
	ConnID string `query:"connId" binding:"required"`
}

// handleGetWorkspace 读取某连接的查询工作区。
func handleGetWorkspace(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req WorkspaceGetReq) (service.WorkspaceState, error) {
		state, _ := svc.LoadWorkspace(req.ConnID)
		if state.Tabs == nil {
			state.Tabs = []service.WorkspaceTab{}
		}
		return state, nil
	})
}

// WorkspaceSaveReq 查询工作区保存请求（整体覆盖）
type WorkspaceSaveReq struct {
	ConnID   string                 `json:"connId" binding:"required"`
	Tabs     []service.WorkspaceTab `json:"tabs" binding:"required"`
	ActiveID string                 `json:"activeId"`
}

// handleSaveWorkspace 保存某连接的查询工作区（整体覆盖）。
func handleSaveWorkspace(svc *service.Service) gin.HandlerFunc {
	return cygin.Handle(func(c *gin.Context, req WorkspaceSaveReq) (map[string]any, error) {
		state := service.WorkspaceState{Tabs: req.Tabs, ActiveID: req.ActiveID}
		if err := svc.SaveWorkspace(req.ConnID, state); err != nil {
			return nil, renderErr(c, err)
		}
		return map[string]any{"ok": true}, nil
	})
}
